From 99bd72e39a7f578aa0e50e1f6e3a8a2257e3b95a Mon Sep 17 00:00:00 2001
From: Gohar Irfan Chaudhry <girfan@mit.edu>
Date: Tue, 8 Nov 2022 14:39:42 -0500
Subject: [PATCH 1/4] no-op pthread_mutex_destroy to let libc exit handlers
 succeed

---
 shim/sync.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/shim/sync.c b/shim/sync.c
index 4dd0880..b23dd96 100644
--- a/shim/sync.c
+++ b/shim/sync.c
@@ -82,7 +82,7 @@ int pthread_mutex_unlock(pthread_mutex_t *mutex)
 
 int pthread_mutex_destroy(pthread_mutex_t *mutex)
 {
-	NOTSELF(pthread_mutex_destroy, mutex);
+//	NOTSELF(pthread_mutex_destroy, mutex);
 	return 0;
 }
 
-- 
2.34.1

