From ba23bebc85b7a7c01d74f889a85e1f7dc59f77dd Mon Sep 17 00:00:00 2001
From: Gohar Irfan Chaudhry <girfan@mit.edu>
Date: Tue, 8 Nov 2022 14:39:42 -0500
Subject: [PATCH 01/33] no-op pthread_mutex_destroy to let libc exit handlers
 succeed

---
 shim/sync.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/shim/sync.c b/shim/sync.c
index 30053f1e..2bb76b49 100644
--- a/shim/sync.c
+++ b/shim/sync.c
@@ -81,7 +81,7 @@ int pthread_mutex_unlock(pthread_mutex_t *mutex)
 
 int pthread_mutex_destroy(pthread_mutex_t *mutex)
 {
-	NOTSELF(pthread_mutex_destroy, mutex);
+//	NOTSELF(pthread_mutex_destroy, mutex);
 	return 0;
 }
 
-- 
2.39.2

