From 522adac02b8a902a1fae9594327360bfdfdf8418 Mon Sep 17 00:00:00 2001
From: Gohar Irfan Chaudhry <girfan@mit.edu>
Date: Tue, 8 Nov 2022 14:39:42 -0500
Subject: [PATCH] no-op pthread_mutex_destroy to let libc exit handlers succeed

---
 shim/sync.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/shim/sync.c b/shim/sync.c
index c42c31b..a18d3c5 100644
--- a/shim/sync.c
+++ b/shim/sync.c
@@ -82,7 +82,7 @@ int pthread_mutex_unlock(pthread_mutex_t *mutex)
 
 int pthread_mutex_destroy(pthread_mutex_t *mutex)
 {
-	NOTSELF_1ARG(int, __func__, mutex);
+	// NOTSELF_1ARG(int, __func__, mutex);
 	return 0;
 }
 
-- 
2.25.1

