From c4b8f0459571f83649033e3ef336e13af65c2839 Mon Sep 17 00:00:00 2001
From: Josh Fried <joshuafried@gmail.com>
Date: Sun, 20 Nov 2022 21:49:09 -0500
Subject: [PATCH 2/3] runtime: export access to stack pointer for junction

---
 inc/runtime/thread.h |  2 ++
 runtime/sched.c      | 12 ++++++++++++
 2 files changed, 14 insertions(+)

diff --git a/inc/runtime/thread.h b/inc/runtime/thread.h
index d71d30f..2e626a8 100644
--- a/inc/runtime/thread.h
+++ b/inc/runtime/thread.h
@@ -26,6 +26,8 @@ extern void thread_ready(thread_t *thread);
 extern void thread_ready_head(thread_t *thread);
 extern thread_t *thread_create(thread_fn_t fn, void *arg);
 extern thread_t *thread_create_with_buf(thread_fn_t fn, void **buf, size_t len);
+extern uint64_t *get_tf_rsp(thread_t *thread);
+
 
 DECLARE_PERTHREAD(thread_t *, __self);
 DECLARE_PERTHREAD(unsigned int, kthread_idx);
diff --git a/runtime/sched.c b/runtime/sched.c
index 2410edc..f363a37 100644
--- a/runtime/sched.c
+++ b/runtime/sched.c
@@ -851,6 +851,18 @@ thread_t *thread_create_with_buf(thread_fn_t fn, void **buf, size_t buf_len)
 	return th;
 }
 
+
+/**
+ * get_tf_rsp - get a pointer to the stored rsp value in a thread's trapframe
+ * @th: the thread
+ *
+ * Returns a pointer to the stored rsp value
+ */
+uint64_t *get_tf_rsp(thread_t *th)
+{
+	return &th->tf.rsp;
+}
+
 /**
  * thread_spawn - creates and launches a new thread
  * @fn: a function pointer to the starting method of the thread
-- 
2.34.1

