From f68caf493d8a3bb4b2e629c8d77561ffca49966b Mon Sep 17 00:00:00 2001
From: Josh Fried <joshuafried@gmail.com>
Date: Mon, 21 Nov 2022 00:00:45 -0500
Subject: [PATCH 3/3] set config optimize to true

---
 build/config | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/build/config b/build/config
index 023e2e1..963e87b 100644
--- a/build/config
+++ b/build/config
@@ -9,7 +9,7 @@ CONFIG_SPDK=n
 # Enable debug build mode (slower but enables several runtime checks)
 CONFIG_DEBUG=n
 # Enable additional compiler optimizations (may reduce compatibility)
-CONFIG_OPTIMIZE=n
+CONFIG_OPTIMIZE=y
 # Allow runtimes to access Mellanox ConnectX-5 NICs directly (kernel bypass)
 CONFIG_DIRECTPATH=n
 # Build with clang instead of gcc
-- 
2.34.1

