From f3361dea4be65ba856fae3bc8199217a62b40bb2 Mon Sep 17 00:00:00 2001
From: Josh Fried <joshuafried@gmail.com>
Date: Sun, 19 Feb 2023 00:34:47 -0500
Subject: [PATCH 14/35] add space for Thread class in struct thread

---
 inc/runtime/thread.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/inc/runtime/thread.h b/inc/runtime/thread.h
index 9985884..453bd29 100644
--- a/inc/runtime/thread.h
+++ b/inc/runtime/thread.h
@@ -119,6 +119,7 @@ struct thread {
      // Trapframe used by junction to stash registers on syscall entry
     struct thread_tf	junction_tf;
     void 		*xsave_area;
+    unsigned long    junction_tstate_buf[8];
 #ifdef GC
     struct list_node    gc_link;
     unsigned int        onk;
-- 
2.43.0

