From 8c4d58d14032bac69befdc4bd8c45b5f9c6e0798 Mon Sep 17 00:00:00 2001
From: "girfan@mit.edu" <girfan@mit.edu>
Date: Tue, 31 Jan 2023 13:12:33 -0500
Subject: [PATCH 14/15] Do not call mbind in __mem_map_anom

---
 base/mem.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/base/mem.c b/base/mem.c
index 8e1314a..e7d9b8b 100644
--- a/base/mem.c
+++ b/base/mem.c
@@ -83,15 +83,17 @@ __mem_map_anom(void *base, size_t len, size_t pgsize,
 	if ((intptr_t)addr < 0)
 		return MAP_FAILED;
 
+  /*
 	BUILD_ASSERT(sizeof(unsigned long) * 8 >= NNUMA);
 	if (syscall_mbind(addr, len, numa_policy, mask ? mask : NULL,
 		  mask ? NNUMA + 1 : 0, MPOL_MF_STRICT | MPOL_MF_MOVE))
 		goto fail;
+  */
 
 	touch_mapping(addr, len, pgsize);
 	return addr;
 
-fail:
+// fail:
 	munmap(addr, len);
 	return MAP_FAILED;
 }
-- 
2.37.2

