From 6eefe3dbf400c5e2fb80eefc56dda9c7572b8c7a Mon Sep 17 00:00:00 2001
From: Josh Fried <joshuafried@gmail.com>
Date: Fri, 17 Feb 2023 17:01:54 -0500
Subject: [PATCH 19/19] tcp listen: allow localhost addr

---
 runtime/net/tcp.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/runtime/net/tcp.c b/runtime/net/tcp.c
index 199037a..846cb3a 100644
--- a/runtime/net/tcp.c
+++ b/runtime/net/tcp.c
@@ -471,7 +471,7 @@ int tcp_listen(struct netaddr laddr, int backlog, tcpqueue_t **q_out)
 		return -EINVAL;
 
 	/* only can support one local IP so far */
-	if (laddr.ip == 0)
+	if (laddr.ip == 0 || laddr.ip == MAKE_IP_ADDR(127, 0, 0, 1))
 		laddr.ip = netcfg.addr;
 	else if (laddr.ip != netcfg.addr)
 		return -EINVAL;
-- 
2.37.2

