From dc4f49a4685c1e9c1179e7b56847a357b07dc470 Mon Sep 17 00:00:00 2001
From: Josh Fried <joshuafried@gmail.com>
Date: Sun, 13 Aug 2023 16:56:31 +0000
Subject: [PATCH 19/35] disable stack protector

---
 build/shared.mk | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/build/shared.mk b/build/shared.mk
index 73c6c28a..f742b16b 100644
--- a/build/shared.mk
+++ b/build/shared.mk
@@ -9,7 +9,7 @@ include $(ROOT_PATH)/build/config
 
 # shared toolchain definitions
 INC = -I$(ROOT_PATH)/inc
-FLAGS  = -g -Wall -D_GNU_SOURCE $(INC) -m64 -mxsavec -m64 -mxsave -m64 -mfsgsbase
+FLAGS  = -g -Wall -D_GNU_SOURCE $(INC) -m64 -mxsavec -m64 -mxsave -m64 -mfsgsbase -fno-stack-protector
 
 ifeq ($(CONFIG_NO_UINTR),n)
 FLAGS += -muintr -DCONFIG_UINTR
-- 
2.43.0

