From 5cef5e7d4f406df23a053197f733c12bc536fbbc Mon Sep 17 00:00:00 2001
From: Josh Fried <joshuafried@gmail.com>
Date: Tue, 20 Jun 2023 15:11:37 -0400
Subject: [PATCH 25/28] save some memory

---
 iokernel/control.c | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/iokernel/control.c b/iokernel/control.c
index 1413a42..17f8360 100644
--- a/iokernel/control.c
+++ b/iokernel/control.c
@@ -205,7 +205,6 @@ static struct proc *control_create_proc(mem_key_t key, size_t len,
 	size_t nr_pages;
 	struct proc *p = NULL;
 	struct thread_spec *threads = NULL;
-	unsigned long *overflow_queue = NULL;
 	void *shbuf = NULL;
 	int i, ret;
 
@@ -311,9 +310,11 @@ static struct proc *control_create_proc(mem_key_t key, size_t len,
 
 	p->max_overflows = hdr.egress_buf_count;
 	p->nr_overflows = 0;
-	p->overflow_queue = overflow_queue = malloc(sizeof(unsigned long) * p->max_overflows);
-	if (overflow_queue == NULL)
-		goto fail;
+	if (!cfg.vfio_directpath) {
+		p->overflow_queue = malloc(sizeof(unsigned long) * p->max_overflows);
+		if (p->overflow_queue == NULL)
+			goto fail;
+	}
 
 	/* free temporary allocations */
 	free(threads);
@@ -321,7 +322,8 @@ static struct proc *control_create_proc(mem_key_t key, size_t len,
 	return p;
 
 fail:
-	free(overflow_queue);
+	if (p)
+		free(p->overflow_queue);
 	free(threads);
 	free(p);
 	if (reg.base)
-- 
2.39.2

