From 46442a7ec8d747125dd67f0b726e2baeff2a9552 Mon Sep 17 00:00:00 2001
From: Josh Fried <joshuafried@gmail.com>
Date: Sun, 14 Jul 2024 21:07:42 -0400
Subject: [PATCH 31/35] thread buffer tweaks

---
 inc/runtime/thread.h | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/inc/runtime/thread.h b/inc/runtime/thread.h
index 09f38beb..71b1ad39 100644
--- a/inc/runtime/thread.h
+++ b/inc/runtime/thread.h
@@ -68,7 +68,7 @@ struct thread {
     bool        xsave_area_in_use:1;
     atomic8_t        interrupt_state;
     struct thread_tf    *entry_regs;
-    unsigned long    junction_tstate_buf[25];
+    unsigned long    junction_tstate_buf[8];
     struct stack        *stack;
     uint16_t        last_cpu;
     uint16_t        cur_kthread;
@@ -82,6 +82,7 @@ struct thread {
     struct list_node    gc_link;
     unsigned int        onk;
 #endif
+    unsigned long    junction_cold_state_buf[32];
 };
 
 extern uint64_t thread_get_total_cycles(thread_t *th);
-- 
2.43.0

