From ad43ba9d20726d5258db6e18966920880a6b2a49 Mon Sep 17 00:00:00 2001
From: Josh Fried <joshuafried@gmail.com>
Date: Sun, 26 May 2024 16:39:39 -0400
Subject: [PATCH 35/35] assert header fix

---
 inc/base/assert.h | 44 +++++++++++++++++++++++---------------------
 1 file changed, 23 insertions(+), 21 deletions(-)

diff --git a/inc/base/assert.h b/inc/base/assert.h
index fdc154b..61deda5 100644
--- a/inc/base/assert.h
+++ b/inc/base/assert.h
@@ -2,7 +2,8 @@
  * assert.h - support for assertions
  */
 
-#pragma once
+#ifndef __BASE_ASSERT_H
+#define __BASE_ASSERT_H
 
 #include <base/stddef.h>
 
@@ -19,26 +20,6 @@ extern void logk_bug(bool fatal, const char *expr,
 #define __build_assert_if_constant(cond)
 #endif /* __CHECKER__ */
 
-#undef assert
-/* these assertions will get compiled out in release builds (fails on false) */
-#if DEBUG
-#define assert(cond)						\
-        do {							\
-		__build_assert_if_constant(cond);		\
-		if (unlikely(!(cond))) {			\
-			logk_bug(true, __cstr(cond),		\
-				 __FILE__, __LINE__, __func__);	\
-			__builtin_unreachable();		\
-		}						\
-        } while (0)
-#else /* DEBUG */
-#define assert(cond)						\
-	do {							\
-		__build_assert_if_constant(cond);		\
-		(void)sizeof(cond);				\
-	} while (0)
-#endif /* DEBUG */
-
 /**
  * BUG - a fatal code-path that doesn't compile out in release builds
  */
@@ -129,3 +110,24 @@ extern void logk_bug(bool fatal, const char *expr,
 #else /* __CHECKER__ */
 #define BUILD_ASSERT_MSG(cond, msg)
 #endif /* __CHECKER__ */
+
+#endif // __BASE_ASSERT_H
+#undef assert
+/* these assertions will get compiled out in release builds (fails on false) */
+#if DEBUG
+#define assert(cond)						\
+        do {							\
+		__build_assert_if_constant(cond);		\
+		if (unlikely(!(cond))) {			\
+			logk_bug(true, __cstr(cond),		\
+				 __FILE__, __LINE__, __func__);	\
+			__builtin_unreachable();		\
+		}						\
+        } while (0)
+#else /* DEBUG */
+#define assert(cond)						\
+	do {							\
+		__build_assert_if_constant(cond);		\
+		(void)sizeof(cond);				\
+	} while (0)
+#endif /* DEBUG */
-- 
2.43.0

