From 4ee5bd7b930aebaec30dc3f23e1b12e6b718fb8e Mon Sep 17 00:00:00 2001
From: Josh Fried <joshuafried@gmail.com>
Date: Mon, 17 Feb 2025 03:34:38 +0000
Subject: [PATCH 36/36] build: rebuild when config parameters change

---
 Makefile | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/Makefile b/Makefile
index a5341f1b..665b82c3 100644
--- a/Makefile
+++ b/Makefile
@@ -80,6 +80,8 @@ asm = $(runtime_asm) $(base_asm)
 obj = $(src:.c=.o) $(asm:.S=.o)
 dep = $(obj:.o=.d)
 
+$(obj): build/config Makefile build/shared.mk
+
 ifneq ($(MAKECMDGOALS),clean)
 -include $(dep)   # include all dep files in the makefile
 endif
-- 
2.43.0

