From eb1e021d9774297f922c2d4d5a5cd7b1c42244d6 Mon Sep 17 00:00:00 2001
From: Josh Fried <joshuafried@gmail.com>
Date: Thu, 11 Jul 2024 15:23:11 +0000
Subject: [PATCH 39/41] nit

---
 iokernel/sched.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/iokernel/sched.c b/iokernel/sched.c
index de2b8fa2..3b67b925 100644
--- a/iokernel/sched.c
+++ b/iokernel/sched.c
@@ -994,13 +994,14 @@ int sched_init(void)
 	 */
 
 	for (i = 0; i < cpu_count; i++) {
-		if (cpu_info_tbl[i].package != managed_numa_node && sched_ops != &numa_ops)
+		if (managed_numa_node != -1 &&
+		    cpu_info_tbl[i].package != managed_numa_node &&
+		    sched_ops != &numa_ops)
 			continue;
 
 		if (allowed_cores_supplied &&
 		    !bitmap_test(input_allowed_cores, i))
 			continue;
-
 		bitmap_set(sched_allowed_cores, i);
 	}
 	/* check for minimum number of cores required */
-- 
2.43.0

