From 31e9adb555adfd4ba8f8e34d2fa7f5cf005c2049 Mon Sep 17 00:00:00 2001
From: Josh Fried <joshuafried@gmail.com>
Date: Sun, 23 Feb 2025 15:19:30 -0500
Subject: [PATCH 11/11] getcontext

---
 sysdeps/unix/sysv/linux/x86_64/getcontext.S | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/sysdeps/unix/sysv/linux/x86_64/getcontext.S b/sysdeps/unix/sysv/linux/x86_64/getcontext.S
index e588a8cc07..e6176200f6 100644
--- a/sysdeps/unix/sysv/linux/x86_64/getcontext.S
+++ b/sysdeps/unix/sysv/linux/x86_64/getcontext.S
@@ -102,9 +102,11 @@ L(no_shstk):
 #else
 	movl	$SIG_BLOCK, %edi
 #endif
-	movl	$_NSIG8,%r10d
-	movl	$__NR_rt_sigprocmask, %eax
-	syscall
+	movq	$_NSIG8,%rcx
+	subq $8, %rsp
+	pushq	$__NR_rt_sigprocmask
+	JCALL_NOPRESERVE
+	addq $16, %rsp
 	cmpq	$-4095, %rax		/* Check %rax for error.  */
 	jae	SYSCALL_ERROR_LABEL	/* Jump to error handler if error.  */
 
-- 
2.43.0

